### Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.



In [1]:
f = open('horse_data.names', 'r', encoding = 'utf8')

In [2]:
many =f.readlines()

In [3]:
info = many[29:264]
info

['  1:  surgery?\n',
 '          1 = Yes, it had surgery\n',
 '          2 = It was treated without surgery\n',
 '\n',
 '  2:  Age\n',
 '          1 = Adult horse\n',
 '          2 = Young (< 6 months)\n',
 '\n',
 '  3:  Hospital Number\n',
 '          - numeric id\n',
 '          - the case number assigned to the horse\n',
 '            (may not be unique if the horse is treated > 1 time)\n',
 '\n',
 '  4:  rectal temperature\n',
 '          - linear\n',
 '          - in degrees celsius.\n',
 '          - An elevated temp may occur due to infection.\n',
 '          - temperature may be reduced when the animal is in late shock\n',
 '          - normal temp is 37.8\n',
 '          - this parameter will usually change as the problem progresses\n',
 '               eg. may start out normal, then become elevated because of\n',
 '                   the lesion, passing back through the normal range as the\n',
 '                   horse goes into shock\n',
 '  5:  pulse\n',
 '          - line

### КОММЕНТ
Здесь в верхней части оставил нужную инфо в переменной info, 
ниже я перевел в json строки чтобы скопировать нужные названия колонок через REGEX

In [4]:
import json
y = json.dumps(info)
y

'["  1:  surgery?\\n", "          1 = Yes, it had surgery\\n", "          2 = It was treated without surgery\\n", "\\n", "  2:  Age\\n", "          1 = Adult horse\\n", "          2 = Young (< 6 months)\\n", "\\n", "  3:  Hospital Number\\n", "          - numeric id\\n", "          - the case number assigned to the horse\\n", "            (may not be unique if the horse is treated > 1 time)\\n", "\\n", "  4:  rectal temperature\\n", "          - linear\\n", "          - in degrees celsius.\\n", "          - An elevated temp may occur due to infection.\\n", "          - temperature may be reduced when the animal is in late shock\\n", "          - normal temp is 37.8\\n", "          - this parameter will usually change as the problem progresses\\n", "               eg. may start out normal, then become elevated because of\\n", "                   the lesion, passing back through the normal range as the\\n", "                   horse goes into shock\\n", "  5:  pulse\\n", "          - lin

In [5]:
import re
find_columns = re.findall(r'\d{1,2}\:\s+\w+', y)
print(find_columns)

['1:  surgery', '2:  Age', '3:  Hospital', '4:  rectal', '5:  pulse', '6:  respiratory', '7:  temperature', '8:  peripheral', '9:  mucous', '10: capillary', '11: pain', '12: peristalsis', '13: abdominal', '14: nasogastric', '15: nasogastric', '16: nasogastric', '17: rectal', '18: abdomen', '19: packed', '20: total', '21: abdominocentesis', '22: abdomcentesis', '23: outcome', '24: surgical', '27: type', '28: cp_data']


In [6]:
import pandas as pd


In [7]:
horse_data = pd.read_csv('horse_data.csv')

In [8]:
horse_data = pd.DataFrame(horse_data)

In [9]:
horse_data.columns =['surgery?', 'age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities'
                    , 'peripheral pulse', 'mucous membranes', 'capillary refill time', 'pain', 'peristalsis','abdominal distension',
                    'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination - feces', 'abdomen',
                     'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome',
                     'surgical lesion', 'type of lesion', 'type of lesion', 'type of lesion', 'cp_data' 
                    ]

In [10]:
horse_data = horse_data.replace(to_replace='?', value='NaN')
horse_data

surgery?  age  Hospital Number rectal temperature pulse respiratory rate  \
0          1    1           534817               39.2    88               20   
1          2    1           530334              38.30    40               24   
2          1    9          5290409              39.10   164               84   
3          2    1           530255              37.30   104               35   
4          2    1           528355                NaN   NaN              NaN   
..       ...  ...              ...                ...   ...              ...   
294        1    1           533886                NaN   120               70   
295        2    1           527702              37.20    72               24   
296        1    1           529386              37.50    72               30   
297        1    1           530612              36.50   100               24   
298        1    1           534618               37.2    40               20   

    temperature of extremities peripheral pulse mucous membranes  \
0                          NaN              NaN                4   
1                            1                1                3   
2                            4                1                6   
3                          NaN              NaN                6   
4                            2                1                3   
..                         ...              ...              ...   
294                          4              NaN                4   
295                          3                2                4   
296                          4                3                4   
297                          3                3                3   
298                        NaN              NaN              NaN   

    capillary refill time  ... packed cell volume total protein  \
0                       1  ...                 50            85   
1                       1  ...              33.00          6.70   
2                       2  ...              48.00          7.20   
3                       2  ...              74.00          7.40   
4                       1  ...                NaN           NaN   
..                    ...  ...                ...           ...   
294                     2  ...                 55            65   
295                     2  ...              44.00           NaN   
296                     1  ...              60.00          6.80   
297                     1  ...              50.00          6.00   
298                   NaN  ...                 36            62   

    abdominocentesis appearance abdomcentesis total protein outcome  \
0                             2                           2       3   
1                           NaN                         NaN       1   
2                             3                        5.30       2   
3                           NaN                         NaN       2   
4                           NaN                         NaN       1   
..                          ...                         ...     ...   
294                         NaN                         NaN       3   
295                           3                        3.30       3   
296                         NaN                         NaN       2   
297                           3                        3.40       1   
298                           1                           1       3   

    surgical lesion type of lesion type of lesion type of lesion cp_data  
0                 2           2208              0              0       2  
1                 2              0              0              0       1  
2                 1           2208              0              0       1  
3                 2           4300              0              0       2  
4                 2              0              0              0       2  
..              ...            ...            ...            ...     ...  
294               2           3205         

In [11]:
sorted_horse_data = horse_data[['surgery?', 'age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 
           'outcome']]
sorted_horse_data

surgery?  age rectal temperature pulse respiratory rate  \
0          1    1               39.2    88               20   
1          2    1              38.30    40               24   
2          1    9              39.10   164               84   
3          2    1              37.30   104               35   
4          2    1                NaN   NaN              NaN   
..       ...  ...                ...   ...              ...   
294        1    1                NaN   120               70   
295        2    1              37.20    72               24   
296        1    1              37.50    72               30   
297        1    1              36.50   100               24   
298        1    1               37.2    40               20   

    temperature of extremities pain outcome  
0                          NaN    3       3  
1                            1    3       1  
2                            4    2       2  
3                          NaN  NaN       2  
4                            2    2       1  
..                         ...  ...     ...  
294                          4    2       3  
295                          3    4       3  
296                          4    4       2  
297                          3    3       1  
298                        NaN  NaN       3  

[299 rows x 8 columns]

### Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.



In [12]:
sorted_horse_data.head(50)

surgery?  age rectal temperature pulse respiratory rate  \
0         1    1               39.2    88               20   
1         2    1              38.30    40               24   
2         1    9              39.10   164               84   
3         2    1              37.30   104               35   
4         2    1                NaN   NaN              NaN   
5         1    1              37.90    48               16   
6         1    1                NaN    60              NaN   
7         2    1                NaN    80               36   
8         2    9              38.30    90              NaN   
9         1    1              38.10    66               12   
10        2    1              39.10    72               52   
11        1    1              37.20    42               12   
12        2    9              38.00    92               28   
13        1    1               38.2    76               28   
14        1    1              37.60    96               48   
15        1    9                NaN   128               36   
16        2    1              37.50    48               24   
17        1    1              37.60    64               21   
18        2    1               39.4   110               35   
19        1    1              39.90    72               60   
20        2    1              38.40    48               16   
21        1    1              38.60    42               34   
22        1    9               38.3   130               60   
23        1    1               38.1    60               12   
24        2    1              37.80    60               42   
25        1    1              38.30    72               30   
26        1    1              37.80    48               12   
27        1    1                NaN   NaN              NaN   
28        2    1               37.7    48              NaN   
29        2    1              37.70    96               30   
30        2    1              37.20   108               12   
31        1    1              37.20    60              NaN   
32        1    1              38.20    64               28   
33        1    1                NaN   100               30   
34        2    1                NaN   104               24   
35        2    1              38.30   112               16   
36        1    1               37.8    72              NaN   
37        2    1              38.60    52              NaN   
38        1    9              39.20   146               96   
39        1    1                NaN    88              NaN   
40        2    9              39.00   150               72   
41        2    1              38.00    60               12   
42        1    1                NaN   120              NaN   
43        1    1               35.4   140               24   
44        2    1                NaN   120              NaN   
45        1    1              37.90    60               15   
46        2    1              37.50    48               16   
47        1    1              38.90    80               44   
48        2    1              37.20    84               48   
49        2    1              38.60    46              NaN   

   temperature of extremities pain outcome  
0                         NaN    3       3  
1                           1    3       1  
2                           4    2       2  
3                         NaN  NaN       2  
4                           2    2       1  
5                           1    3       1  
6                           3  NaN       2  
7                           3    4       3  
8                           1    5       1  
9                           3    3       1  
10                          2    2       1  
11                          2    3       1  
12                          1    1       2  
13                          3    3       1  
14                          3    5       2  
15                          3    4       2  
16                        NaN  NaN       1  
17                          1 

In [13]:
sorted_horse_data['age'].mean()

1.6421404682274248

In [14]:
(sorted_horse_data.isna().mean()*100).round(2)
#(titanic.isna().mean() * 100).round(2)

surgery?                      0.0
age                           0.0
rectal temperature            0.0
pulse                         0.0
respiratory rate              0.0
temperature of extremities    0.0
pain                          0.0
outcome                       0.0
dtype: float64

In [15]:
sorted_horse_data.describe()

age
count  299.00000
mean     1.64214
std      2.17730
min      1.00000
25%      1.00000
50%      1.00000
75%      1.00000
max      9.00000

In [16]:
sorted_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   surgery?                    299 non-null    object
 1   age                         299 non-null    int64 
 2   rectal temperature          299 non-null    object
 3   pulse                       299 non-null    object
 4   respiratory rate            299 non-null    object
 5   temperature of extremities  299 non-null    object
 6   pain                        299 non-null    object
 7   outcome                     299 non-null    object
dtypes: int64(1), object(7)
memory usage: 18.8+ KB


In [17]:
sorted_horse_data = sorted_horse_data.replace(to_replace='NaN', value='0')

In [18]:
#sorted_horse_data.astype('int64', errors='ignore').dtypes

In [19]:
sorted_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   surgery?                    299 non-null    object
 1   age                         299 non-null    int64 
 2   rectal temperature          299 non-null    object
 3   pulse                       299 non-null    object
 4   respiratory rate            299 non-null    object
 5   temperature of extremities  299 non-null    object
 6   pain                        299 non-null    object
 7   outcome                     299 non-null    object
dtypes: int64(1), object(7)
memory usage: 18.8+ KB


In [20]:
sorted_horse_data['surgery?'].mean()

4.054218131144185e+295

In [21]:
sorted_horse_data.describe()

age
count  299.00000
mean     1.64214
std      2.17730
min      1.00000
25%      1.00000
50%      1.00000
75%      1.00000
max      9.00000

In [22]:
sorted_horse_data['surgery?'].fillna(sorted_horse_data['surgery?'].mean(), inplace=True)
sorted_horse_data['surgery?'].fillna(sorted_horse_data['surgery?'].mean()).isna().sum()

0

### Значения для столбца "surgery?"

In [23]:
print('mean ', pd.to_numeric(sorted_horse_data['surgery?'], errors='coerce').mean())
print('median', pd.to_numeric(sorted_horse_data['surgery?'], errors='coerce').median())
print('mode', pd.to_numeric(sorted_horse_data['surgery?'], errors='coerce').mode()[0])
print('std', pd.to_numeric(sorted_horse_data['surgery?'], errors='coerce').std())
print('var', pd.to_numeric(sorted_horse_data['surgery?'], errors='coerce').var())

mean  1.391304347826087
median 1.0
mode 1
std 0.495677257874841
var 0.24569594397432162


In [24]:
def described(horses):
    for b in horses:
        print('mean ', b, pd.to_numeric(horses[b], errors='coerce').mean())
        print('median', b, pd.to_numeric(horses[b], errors='coerce').median())
        print('mode', b, pd.to_numeric(horses[b], errors='coerce').mode()[0])
        print('std', b, pd.to_numeric(horses[b], errors='coerce').std())
        print('var', b, pd.to_numeric(horses[b], errors='coerce').var())
    

In [25]:
described(sorted_horse_data)

mean  surgery? 1.391304347826087
median surgery? 1.0
mode surgery? 1
std surgery? 0.495677257874841
var surgery? 0.24569594397432162
mean  age 1.6421404682274248
median age 1.0
mode age 1
std age 2.1772997793587714
var age 4.740634329195755
mean  rectal temperature 30.50769230769233
median rectal temperature 38.0
mode rectal temperature 0.0
std rectal temperature 15.325376190976206
var rectal temperature 234.86715539494037
mean  pulse 66.16053511705685
median pulse 60.0
mode pulse 48
std pulse 33.75823510340228
var pulse 1139.618437296582
mean  respiratory rate 24.525083612040135
median respiratory rate 22.0
mode respiratory rate 0
std respiratory rate 19.923221912773116
var respiratory rate 396.9347713856028
mean  temperature of extremities 1.9063545150501673
median temperature of extremities 2.0
mode temperature of extremities 3
std temperature of extremities 1.3150878787863944
var temperature of extremities 1.7294561289308983
mean  pain 2.4013377926421406
median pain 2.0
mode pain 3

In [26]:
for b in sorted_horse_data:
    print('mean ', b, pd.to_numeric(sorted_horse_data[b], errors='coerce').mean())
    print('median', b, pd.to_numeric(sorted_horse_data[b], errors='coerce').median())
    print('mode', b, pd.to_numeric(sorted_horse_data[b], errors='coerce').mode()[0])
    print('std', b, pd.to_numeric(sorted_horse_data[b], errors='coerce').std())
    print('var', b, pd.to_numeric(sorted_horse_data[b], errors='coerce').var())

mean  surgery? 1.391304347826087
median surgery? 1.0
mode surgery? 1
std surgery? 0.495677257874841
var surgery? 0.24569594397432162
mean  age 1.6421404682274248
median age 1.0
mode age 1
std age 2.1772997793587714
var age 4.740634329195755
mean  rectal temperature 30.50769230769233
median rectal temperature 38.0
mode rectal temperature 0.0
std rectal temperature 15.325376190976206
var rectal temperature 234.86715539494037
mean  pulse 66.16053511705685
median pulse 60.0
mode pulse 48
std pulse 33.75823510340228
var pulse 1139.618437296582
mean  respiratory rate 24.525083612040135
median respiratory rate 22.0
mode respiratory rate 0
std respiratory rate 19.923221912773116
var respiratory rate 396.9347713856028
mean  temperature of extremities 1.9063545150501673
median temperature of extremities 2.0
mode temperature of extremities 3
std temperature of extremities 1.3150878787863944
var temperature of extremities 1.7294561289308983
mean  pain 2.4013377926421406
median pain 2.0
mode pain 3

### Среднее для столбца "pulse"

In [27]:
pd.to_numeric(sorted_horse_data['pulse'], errors='coerce').mean()

66.16053511705685

### Среднее для столбца "rectal temperature"

In [28]:
pd.to_numeric(sorted_horse_data['rectal temperature'], errors='coerce').mean()

30.50769230769233

###  Среднее для остальных стлбцов

In [43]:
print('respiratory rate - ',pd.to_numeric(sorted_horse_data['respiratory rate'], errors='coerce').mean())
print('temperature of extremities - ', pd.to_numeric(sorted_horse_data['temperature of extremities'], errors='coerce').mean())
print('pain - ', pd.to_numeric(sorted_horse_data['pain'], errors='coerce').mean())
print('outcome - ', pd.to_numeric(sorted_horse_data['outcome'], errors='coerce').mean())

respiratory rate -  24.525083612040135
temperature of extremities -  1.9063545150501673
pain -  2.4013377926421406
outcome -  1.5451505016722409


In [44]:
import numpy as np

In [45]:
sorted_h_d = sorted_horse_data.describe(include='all')
sorted_h_d = sorted_h_d.replace(to_replace='?', value='NaN')
sorted_h_d

surgery?        age rectal temperature pulse respiratory rate  \
count       299  299.00000                299   299              299   
unique        3        NaN                 66    53               41   
top           1        NaN                  0    48                0   
freq        180        NaN                 60    28               58   
mean        NaN    1.64214                NaN   NaN              NaN   
std         NaN    2.17730                NaN   NaN              NaN   
min         NaN    1.00000                NaN   NaN              NaN   
25%         NaN    1.00000                NaN   NaN              NaN   
50%         NaN    1.00000                NaN   NaN              NaN   
75%         NaN    1.00000                NaN   NaN              NaN   
max         NaN    9.00000                NaN   NaN              NaN   

       temperature of extremities pain outcome  
count                         299  299     299  
unique                          5    6       4  
top                             3    3       1  
freq                          108   67     178  
mean                          NaN  NaN     NaN  
std                           NaN  NaN     NaN  
min                           NaN  NaN     NaN  
25%                           NaN  NaN     NaN  
50%                           NaN  NaN     NaN  
75%                           NaN  NaN     NaN  
max                           NaN  NaN     NaN

In [46]:
sorted_h_d.max() 

age    299.0
dtype: float64

### Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [47]:
sorted_h_d.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11 entries, count to max
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    4 non-null      object 
 1   age                         8 non-null      float64
 2   rectal temperature          4 non-null      object 
 3   pulse                       4 non-null      object 
 4   respiratory rate            4 non-null      object 
 5   temperature of extremities  4 non-null      object 
 6   pain                        4 non-null      object 
 7   outcome                     4 non-null      object 
dtypes: float64(1), object(7)
memory usage: 792.0+ bytes


### Считаем пропуска в каждом столбце

In [48]:
print(sorted_h_d.isna().sum())

surgery?                      7
age                           3
rectal temperature            7
pulse                         7
respiratory rate              7
temperature of extremities    7
pain                          7
outcome                       7
dtype: int64


###  Каждый столбец имеет 299 строк
### 7/299 = 2.3 %  Думаю можно просто удалить пустые строки

In [54]:
for b in sorted_h_d:
    print('mean ', b, round(pd.to_numeric(sorted_horse_data[b], errors='coerce').mean(), 4))

mean  surgery? 1.3913
mean  age 1.6421
mean  rectal temperature 30.5077
mean  pulse 66.1605
mean  respiratory rate 24.5251
mean  temperature of extremities 1.9064
mean  pain 2.4013
mean  outcome 1.5452


In [55]:
sorted_h_d = sorted_h_d.dropna(axis=0, how='any')
sorted_h_d.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1 entries, count to count
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    1 non-null      object 
 1   age                         1 non-null      float64
 2   rectal temperature          1 non-null      object 
 3   pulse                       1 non-null      object 
 4   respiratory rate            1 non-null      object 
 5   temperature of extremities  1 non-null      object 
 6   pain                        1 non-null      object 
 7   outcome                     1 non-null      object 
dtypes: float64(1), object(7)
memory usage: 72.0+ bytes


In [56]:
print(sorted_h_d.isna().sum())

surgery?                      0
age                           0
rectal temperature            0
pulse                         0
respiratory rate              0
temperature of extremities    0
pain                          0
outcome                       0
dtype: int64


In [58]:
for b in sorted_h_d:
    print('mean ', b, round(pd.to_numeric(sorted_horse_data[b], errors='coerce').mean(), 4))

mean  surgery? 1.3913
mean  age 1.6421
mean  rectal temperature 30.5077
mean  pulse 66.1605
mean  respiratory rate 24.5251
mean  temperature of extremities 1.9064
mean  pain 2.4013
mean  outcome 1.5452


### Комментарий
До удаления пустых строк и после удаления средние значения остались одинаковыми, т.е. удаление строки с пустым значением в одной колонке не повлияло на соседние колонки